File: lib/herb/position.rb
# frozen_string_literal: true

module Herb
  class Position
    attr_reader :line, :column

    def initialize(line, column)
      @line = line
      @column = column
    end

    def self.[](...)
      new(...)
    end

    def self.from(...)
      new(...)
    end

    def to_hash
      { line: line, column: column }
    end

    def to_json(*args)
      to_hash.to_json(*args)
    end

    def tree_inspect
      "(#{line}:#{column})"
    end

    def inspect
      %(#<Herb::Position #{tree_inspect}>)
    end
  end
end

File: javascript/packages/core/src/position.ts
export type SerializedPosition = {
  line: number
  column: number
}

export class Position {
  readonly line: number
  readonly column: number

  static from(position: SerializedPosition): Position
  static from(line: number, column: number): Position
  static from(positionOrLine: SerializedPosition | number, column?: number): Position {
    if (typeof positionOrLine === "number") {
      return new Position(positionOrLine, column!)
    } else {
      return new Position(positionOrLine.line, positionOrLine.column)
    }
  }

  static get zero() {
    return new Position(0, 0)
  }

  constructor(line: number, column: number) {
    this.line = line
    this.column = column
  }

  toHash(): SerializedPosition {
    return { line: this.line, column: this.column }
  }

  toJSON(): SerializedPosition {
    return this.toHash()
  }

  treeInspect(): string {
    return `(${this.line}:${this.column})`
  }

  inspect(): string {
    return `#<Herb::Position ${this.treeInspect()}>`
  }

  toString(): string {
    return this.inspect()
  }
}
